### ISSUE TITLE
loss 未收敛/收敛慢

### ISSUE THREAD

**Issue by fh1999** [Xiaojiu-z/EasyControl#83]
Hi～ 我使用了两个lora 同时训练，输入分辨率512，两个lora输入都是256*256，但是我发现效果非常有限，loss并未像我想象的初期快速下降。

![Image](https://github.com/user-attachments/assets/f109b3a3-0a4d-4414-9088-2b2b4897a189)

我的训练代码是基于x-flux开发的，下面是我适配过去的一些代码，如果你能帮我看看是否有bug，我将万分感谢～：

```python
class MultiSingleStreamBlockLoraProcessor(nn.Module):
    def __init__(self, dim: int, ranks=[], lora_weights=[], network_alphas=[], device=None, dtype=None, cond_width=512, cond_height=512, n_loras=1):
        super().__init__()
        # Initialize a list to store the LoRA layers
        self.n_loras = n_loras
        self.cond_width = cond_width
        self.cond_height = cond_height
        
        # QKV LoRA layers for multiple LoRAs
        self.qkv_loras = nn.ModuleList([
            EasyControlLoRALinearLayer(dim, dim * 3, ranks[i], network_alphas[i], device=device, dtype=dtype, cond_width=cond_width, cond_height=cond_height, number=i, n_loras=n_loras)
            for i in range(n_loras)
        ])
        
        # Projection LoRA layers for multiple LoRAs
        self.proj_loras = nn.ModuleList([
            EasyControlLoRALinearLayer(15360, dim, ranks[i], network_alphas[i], device=device, dtype=dtype, cond_width=cond_width, cond_height=cond_height, number=i, n_loras=n_loras)
            for i in range(n_loras)
        ])
        
        self.lora_weights = lora_weights

    def forward(self, attn: nn.Module, x: Tensor, vec: Tensor, pe: Tensor, cond_hidden_states: Tensor | None = None,cond_vec: Tensor | None = None) -> Tensor | tuple[Tensor, Tensor]:        
        # 处理主要隐藏状态
        mod, _ = attn.modulation(vec)
        x_mod = (1 + mod.scale) * attn.pre_norm(x) + mod.shift

        cond_mod,_ = attn.modulation(cond_vec)
        cond_x_mod = (1 + cond_mod.scale) * attn.pre_norm(cond_hidden_states) + cond_mod.shift

        mod_concat = torch.cat([x_mod, cond_x_mod], dim=-2)

        qkv_concat,mlp_concat = torch.split(attn.linear1(mod_concat), [3 * attn.hidden_size, attn.mlp_hidden_dim], dim=-1)

        # 应用多个 QKV LoRA
        for i in range(self.n_loras):
            qkv_concat = qkv_concat + self.lora_weights[i] * self.qkv_loras[i](mod_concat)

        q, k, v = rearrange(qkv_concat, "B L (K H D) -> K B H L D", K=3, H=attn.num_heads)

        q, k = attn.norm(q, k, v)

        cond_size = self.cond_width // 8 * self.cond_height // 8 * 16 // 64
        block_size =  mod_concat.shape[1] - cond_size * self.n_loras
        scaled_cond_size = cond_size
        scaled_block_size = block_size
        scaled_seq_len = q.shape[2]

        num_cond_blocks = self.n_loras
        mask = torch.ones((scaled_seq_len, scaled_seq_len), device=mod_concat.device)
        mask[ :scaled_block_size, :] = 0  # First block_size row
        for i in range(num_cond_blocks):
            start = i * scaled_cond_size + scaled_block_size
            end = (i + 1) * scaled_cond_size + scaled_block_size
            mask[start:end, start:end] = 0  # Diagonal blocks
        mask = mask * -1e20
        mask = mask.to(q.dtype)

        attn_output = attention(q, k, v, pe=pe, mask=mask)

        cond_hidden_states_attn_output = attn_output[:, block_size:,:]
        vec_attn_output = attn_output[:, : block_size,:]
        
        vec_mlp,cond_mlp = mlp_concat[:,:block_size],mlp_concat[:,block_size:]

        vec_output = attn.linear2(torch.cat((vec_attn_output, attn.mlp_act(vec_mlp)), 2))

        cond_output = attn.linear2(torch.cat((cond_hidden_states_attn_output, attn.mlp_act(cond_mlp)), 2))

        cond_hidden_states = cond_hidden_states + cond_mod.gate * cond_output
        x = x + mod.gate * vec_output

        return x, cond_hidden_states


class MultiDoubleStreamBlockLoraProcessor(nn.Module):
    def __init__(self, dim: int, ranks=[], lora_weights=[], network_alphas=[], device=None, dtype=None, cond_width=512, cond_height=512, n_loras=1):
        super().__init__()
        
        # Initialize a list to store the LoRA layers
        self.n_loras = n_loras
        self.cond_width = cond_width
        self.cond_height = cond_height
        
        # Image QKV LoRA layers for multiple LoRAs
        self.img_qkv_loras = nn.ModuleList([
            EasyControlLoRALinearLayer(dim, dim * 3, ranks[i], network_alphas[i], device=device, dtype=dtype, cond_width=cond_width, cond_height=cond_height, number=i, n_loras=n_loras)
            for i in range(n_loras)
        ])
        
        # Image projection LoRA layers for multiple LoRAs
        self.img_proj_loras = nn.ModuleList([
            EasyControlLoRALinearLayer(dim, dim, ranks[i], network_alphas[i], device=device, dtype=dtype, cond_width=cond_width, cond_height=cond_height, number=i, n_loras=n_loras)
            for i in range(n_loras)
        ])
        
        self.lora_weights = lora_weights

    def forward(self, attn, img, txt, vec, pe, cond_hidden_states: Tensor | None = None,cond_vec: Tensor | None = None, **attention_kwargs):        
        img_mod1, img_mod2 = attn.img_mod(vec)
        txt_mod1, txt_mod2 = attn.txt_mod(vec)

        cond_mod1, cond_mod2 = attn.img_mod(cond_vec)
        
        # prepare image for attention
        img_modulated = attn.img_norm1(img)
        img_modulated = (1 + img_mod1.scale) * img_modulated + img_mod1.shift

        cond_img_modulated = attn.img_norm1(cond_hidden_states)
        cond_img_modulated = (1 + cond_mod1.scale) * cond_img_modulated + cond_mod1.shift

        hidden_states = torch.cat((img_modulated, cond_img_modulated), dim=-2)

        img_qkv = attn.img_attn.qkv(hidden_states)
        
        # 应用多个 image QKV LoRA
        for i in range(self.n_loras):
            img_qkv = img_qkv + self.lora_weights[i] * self.img_qkv_loras[i](hidden_states)
        

        img_q, img_k, img_v = rearrange(img_qkv, "B L (K H D) -> K B H L D", K=3, H=attn.num_heads)
        img_q, img_k = attn.img_attn.norm(img_q, img_k, img_v)

        # prepare txt for attention
        txt_modulated = attn.txt_norm1(txt)
        txt_modulated = (1 + txt_mod1.scale) * txt_modulated + txt_mod1.shift
        txt_qkv = attn.txt_attn.qkv(txt_modulated)
        
        txt_q, txt_k, txt_v = rearrange(txt_qkv, "B L (K H D) -> K B H L D", K=3, H=attn.num_heads)
        txt_q, txt_k = attn.txt_attn.norm(txt_q, txt_k, txt_v)
        

        q = torch.cat((txt_q, img_q), dim=2)
        k = torch.cat((txt_k, img_k), dim=2)
        v = torch.cat((txt_v, img_v), dim=2)
        

        cond_size = self.cond_width // 8 * self.cond_height // 8 * 16 // 64 
        block_size =  hidden_states.shape[1] - cond_size * self.n_loras
        scaled_cond_size = cond_size
        scaled_seq_len = q.shape[2]
        scaled_block_size = scaled_seq_len - cond_size * self.n_loras
        
        num_cond_blocks = self.n_loras
        mask = torch.ones((scaled_seq_len, scaled_seq_len), device=hidden_states.device)
        mask[ :scaled_block_size, :] = 0  # First block_size row
        for i in range(num_cond_blocks):
            start = i * scaled_cond_size + scaled_block_size
            end = (i + 1) * scaled_cond_size + scaled_block_size
            mask[start:end, start:end] = 0  # Diagonal blocks
        mask = mask * -1e20
        mask = mask.to(q.dtype)
        

        attn_output = attention(q, k, v, pe=pe, mask=mask)


        txt_attn, hidden_states = (
            attn_output[:, : txt.shape[1]],
            attn_output[:, txt.shape[1] :],
        )

        hidden_states = attn.img_attn.proj(hidden_states)
        
        for i in range(self.n_loras):
            hidden_states = hidden_states + self.lora_weights[i] * self.img_proj_loras[i](hidden_states)

        
        img_attn,cond_attn = (
            hidden_states[:, :block_size],
            hidden_states[:, block_size:],
        )


        img = img + img_mod1.gate * img_attn
        cond_hidden_states = cond_hidden_states + cond_mod1.gate * cond_attn

        txt = txt + txt_mod1.gate * attn.txt_attn.proj(txt_attn)
        txt = txt + txt_mod2.gate * attn.txt_mlp((1 + txt_mod2.scale) * attn.txt_norm2(txt) + txt_mod2.shift)


        img = img + img_mod2.gate * attn.img_mlp((1 + img_mod2.scale) * attn.img_norm2(img) + img_mod2.shift)
        cond_hidden_states = cond_hidden_states + cond_mod2.gate * attn.img_mlp((1 + cond_mod2.scale) * attn.img_norm2(cond_hidden_states) + cond_mod2.shift)
        return img, txt, cond_hidden_states


```
